src/parsers/pm2ps_parser.jl

```diff
@@ -256,194 +259,233 @@ function make_thermal_gen(gen_name, d, bus)
-    generators["Thermal"] = Dict{String,Any}()
-    generators["Hydro"] = Dict{String,Any}()
-    generators["Renewable"] = Dict{String,Any}()
-    generators["Renewable"]["PV"]= Dict{String,Any}()
-    generators["Renewable"]["RTPV"]= Dict{String,Any}()
-    generators["Renewable"]["WIND"]= Dict{String,Any}()
-    generators["Storage"] = Dict{String,Any}() # not currently used? JJS 3/13/19
-    
     if !haskey(data, "gen")
+        @error "There are no Generators in this file"
         return nothing
     end
-    
-    fuel = []
-    gen_name =[]
-    type_gen =[]
-    for (d_key,d) in data["gen"]
-
-        fuel = uppercase(get(d, "fuel", "generic"))
-        type_gen = uppercase(get(d, "type", "generic"))
-        if haskey(d, "name")
-            gen_name = d["name"]
-        elseif haskey(d, "source_id")
-            gen_name = strip(string(d["source_id"][1])*"-"*d["source_id"][2])
+
+    genmap_file = get(kwargs, :genmap_file, nothing)
+    genmap = get_generator_mapping(genmap_file)
+
+    for (name, pm_gen) in data["gen"]
+        if haskey(pm_gen, "name")
+            gen_name = pm_gen["name"]
+        elseif haskey(pm_gen, "source_id")
+            gen_name = strip(string(pm_gen["source_id"][1]) * "-" * pm_gen["source_id"][2])
         else
-            gen_name = d_key
+            gen_name = name
         end
         
-        bus = find_bus(Buses, d)
-
-        assigned = false 
-        for (rkey, rval) in generators["Renewable"]
-            fuelkeys = genmap_dict[rkey]["fuel"]
-            typekeys = genmap_dict[rkey]["type"]
-            if fuel in fuelkeys && type_gen in typekeys
-                generators["Renewable"][rkey][gen_name] = make_ren_gen(gen_name, d, bus)
-                assigned = true
-                break
-            end
-        end
-        if !assigned
-            fuelkeys = genmap_dict["Hydro"]["fuel"]
-            typekeys = genmap_dict["Hydro"]["type"]
-            if fuel in fuelkeys && type_gen in typekeys
-                generators["Hydro"][gen_name] = make_hydro_gen(gen_name, d, bus)
-                assigned = true
-            end
+        bus = bus_number_to_bus[pm_gen["gen_bus"]]
+        fuel = get(pm_gen, "fuel", "generic")
+        unit_type = get(pm_gen, "type", "generic")
+        @debug "Found generator" gen_name bus fuel unit_type
+
+        gen_type = get_generator_type(fuel, unit_type, genmap)
+        if gen_type == ThermalStandard
+            generator = make_thermal_gen(gen_name, pm_gen, bus)
+        elseif gen_type == HydroDispatch
+            generator = make_hydro_gen(gen_name, pm_gen, bus)
+        elseif gen_type == RenewableDispatch
+            generator = make_renewable_dispatch(gen_name, pm_gen, bus)
+        elseif gen_type == RenewableFix
+            generator = make_renewable_fix(gen_name, pm_gen, bus)
+        elseif gen_type == GenericBattery
+            @warn "Skipping GenericBattery"
+            continue
+            # TODO
```

Comment:
I'm not sure why this wasn't covered before. It should be simple to add.  Let me know if I should fill it in.

Reply:
We probably need to add another Battery type that fits with PM's Storage Data Model

Reply:
I'll leave it for now.